netty/src/main/java/io/grpc/netty/WriteQueue.java

```diff
@@ -120,15 +133,25 @@ ChannelFuture enqueue(Object command, ChannelPromise promise, boolean flush) {
   private void flush() {
     try {
       boolean flushed = false;
+      // We can't just call flush after having completely drained the queue, as new objects might
+      // be added to the queue continuously and so it may never get empty. If we never flushed in
+      // that case we would be guaranteed to OOM.
+      // However, we also don't want to flush too often as flushing invokes expensive system
+      // calls and we want to feed Netty with enough data so as to maximize the data written per
+      // system call.
+      int writesBeforeFlush = min(queue.size(), MAX_WRITES_BEFORE_FLUSH);
+      // Always dequeue in chunks, in order to not acquire the queue's lock too often.
       while (queue.drainTo(writeChunk, DEQUE_CHUNK_SIZE) > 0) {
-        while (writeChunk.size() > 0) {
-          writeChunk.poll().run();
+        writesBeforeFlush -= writeChunk.size();
```

Comment:
@carl-mastrangelo well even when swapping the queues, each `poll()` would still have to acquire a lock. I think that's what @louiscryan had in mind there. I wouldn't expect the lock to be contented, as `offer` and `poll` use two separate locks in the `LinkedBlockingQueue`.

Nevertheless, an interesting idea, which we would have to experiment with. I would prefer to do that in a separate PR, as this one is really focused on the flushing.